File: acquisition/cadastre_role_province.py
#!/usr/bin/env python3
"""Batch PROVINCE du rôle foncier — joint la cadastre CLIPPÉE au rôle MAMH, en masse.

Parcourt les cadastres clippés propres normalized/qc-cadastre-lots/*.geojson (~1102
munis QC) et, pour chacun, télécharge + parse le rôle d'évaluation foncière MAMH,
le joint par matricule (= NO_LOT sans espaces) et dépose un parquet enrichi sous
registry/role-foncier/<slug>.parquet.

IDEMPOTENT & RESUMABLE :
  - skip si registry/role-foncier/<slug>.parquet existe déjà (les 30 z∩m∩p inclus)
  - skip si déjà fait dans le checkpoint /tmp/role_province_progress.json
  - borné par --max-seconds ; relancer pour continuer là où on s'est arrêté

RÉSOLUTION code_geo (sans Overpass, comme le clip) :
  slug cadastre (norm NFD, accents décomposés -> lettre de base) ->
    1. index frontières SDA local (MUS_NM_MUN -> MUS_CO_GEO), matching norm()
    2. fallback index rôle MAMH (fetch_index, nom -> code_geo), matching norm()
  ⚠️ On N'UTILISE PAS resolve_muni()/_slugify() du module role_foncier pour la
     résolution NI pour la clé d'upload : son _slugify fait encode('ascii','ignore')
     qui DÉTRUIT les accents ('Montréal'->'montral', "Baie-D'Urfé"->'baie-d-urf'),
     ce qui casse l'alignement avec le slug cadastre. La clé d'upload est TOUJOURS
     le slug cadastre exact, contrôlée nous-mêmes via boto3.

COMMIT (anti-invention) :
  - lots sans matricule au rôle -> champs role_* = null (jamais inventés)
  - upload parquet sous la clé EXACTE registry/role-foncier/<slug>.parquet
  - log le join% (coverage) par muni

Munis sans rôle (code non résolu / XML 404) -> log no_code / no_role et on continue
(normal : tous les territoires QC n'ont pas de rôle ou de code géo résolvable).

Lecture seule hors registry/role-foncier/.

Usage :
  cadastre_role_province.py [--max-seconds 3000] [--chunk N] [--only SLUG] [--dry-run]
"""
import argparse
import json
import os
import sys
import time
import urllib.error
import urllib.request

sys.path.insert(0, "/home/antoinefa/src/geo")
from acquisition.cadastre_clip_sda import norm
from acquisition.role_foncier import (
    fetch_index,
    parse_role,
    join_lots_role,
)

S3ENV = "/home/antoinefa/src/_acquisition-shared/s3.env"
BUCKET = "sentropic-geo"
PROG = "/tmp/role_province_progress.json"
WORK = "/tmp/role_province"
BOUNDARIES = WORK + "/qc-municipalites.geojson"

CAD_PREFIX = "normalized/qc-cadastre-lots/"
ROLE_PREFIX = "registry/role-foncier/"
SDA_BOUNDARIES_KEY = "normalized/qc-admin-boundaries/qc-municipalites.geojson"

MILLESIME = 2026

# Alias slug cadastre -> code géo (cas où le nom cadastre diffère de la source)
# Aligné sur cadastre_clip_sda.ALIAS_SLUG_TO_CODE.
ALIAS_SLUG_TO_CODE = {
    "eeyou-istchee-james-bay": "99060",
    "hatley-township-municipality": "45043",
}


# ---------------------------------------------------------------------------
# S3
# ---------------------------------------------------------------------------

def s3_client():
    import boto3
    env = {}
    with open(S3ENV) as f:
        for ln in f:
            ln = ln.strip()
            if "=" in ln and not ln.startswith("#"):
                k, v = ln.split("=", 1)
                env[k.strip()] = v.strip()
    return boto3.client(
        "s3", endpoint_url=env["S3_ENDPOINT"],
        region_name=env.get("S3_REGION", "fr-par"),
        aws_access_key_id=env["S3_ACCESS_KEY"],
        aws_secret_access_key=env["S3_SECRET_KEY"],
    )


def exists(s3, key):
    try:
        s3.head_object(Bucket=BUCKET, Key=key)
        return True
    except Exception:
        return False


def list_slugs(s3, prefix, suffix):
    out = []
    pg = s3.get_paginator("list_objects_v2")
    for page in pg.paginate(Bucket=BUCKET, Prefix=prefix):
        for o in page.get("Contents", []):
            k = o["Key"]
            if k.endswith(suffix):
                out.append(k[len(prefix):-len(suffix)])
    return out


# ---------------------------------------------------------------------------
# Checkpoint
# ---------------------------------------------------------------------------

def load_prog():
    try:
        return json.load(open(PROG))
    except Exception:
        return {"done": {}, "skip_exists": [], "no_code": [],
                "no_role": [], "errors": []}


def save_prog(p):
    tmp = PROG + ".tmp"
    json.dump(p, open(tmp, "w"), ensure_ascii=False)
    os.replace(tmp, PROG)


# ---------------------------------------------------------------------------
# Résolution code_geo (slug cadastre -> code géographique)
# ---------------------------------------------------------------------------

class CodeResolver:
    """Résout un slug cadastre (norm NFD) -> code_geo, via SDA puis index rôle MAMH."""

    def __init__(self, boundaries_path):
        # 1) index frontières SDA : norm(MUS_NM_MUN) -> MUS_CO_GEO
        self.by_name_sda = {}
        try:
            g = json.load(open(boundaries_path))
            for f in g.get("features", []):
                p = f.get("properties") or {}
                code = (p.get("MUS_CO_GEO") or p.get("code") or "").strip()
                nm = p.get("MUS_NM_MUN") or p.get("name") or ""
                if code and nm:
                    self.by_name_sda.setdefault(norm(nm), code)
        except Exception as e:
            print("WARN SDA boundaries load failed (%s)" % e)
        # 2) index rôle MAMH : norm(nom) -> code_geo  (fallback)
        self.by_name_role = {}
        try:
            idx = fetch_index(MILLESIME)
            for e in idx.values():
                nm = e.get("nom") or ""
                cg = e.get("code_geo") or ""
                if nm and cg:
                    self.by_name_role.setdefault(norm(nm), cg)
        except Exception as e:
            print("WARN role index fetch failed (%s)" % e)

    def resolve(self, slug):
        """Retourne (code_geo, method) ou (None, 'no-match')."""
        s = norm(slug)
        if s in ALIAS_SLUG_TO_CODE:
            return ALIAS_SLUG_TO_CODE[s], "alias"
        if s in self.by_name_sda:
            return self.by_name_sda[s], "sda-name"
        if s in self.by_name_role:
            return self.by_name_role[s], "role-index"
        # strip progressif des suffixes '--<mrc>' / '--2' (homonymes encodent leur MRC)
        segs = [seg for seg in __import__("re").split(r"-{2,}", slug or "")]
        for i in range(len(segs), 0, -1):
            base = norm("-".join(segs[:i]))
            if base in self.by_name_sda:
                return self.by_name_sda[base], "sda-name-prefix"
            if base in self.by_name_role:
                return self.by_name_role[base], "role-index-prefix"
        return None, "no-match"


# ---------------------------------------------------------------------------
# Rôle : fetch XML par code_geo (404 -> None, sans lever)
# ---------------------------------------------------------------------------

XML_URL_TEMPLATE = (
    "https://donneesouvertes.affmunqc.net/role/RL{code_geo}_{millesime}.xml"
)


def fetch_role_bytes(code_geo, millesime=MILLESIME):
    """Télécharge le XML du rôle pour un code_geo. Retourne bytes ou None si 404."""
    url = XML_URL_TEMPLATE.format(code_geo=code_geo, millesime=millesime)
    try:
        with urllib.request.urlopen(url, timeout=180) as r:
            return r.read()
    except urllib.error.HTTPError as e:
        if e.code in (403, 404):
            return None
        raise


# ---------------------------------------------------------------------------
# Parquet : écrit le GeoJSON enrichi en parquet (properties only, schéma 30)
# ---------------------------------------------------------------------------

def write_parquet(enriched_fc, out_path):
    import pandas as pd
    import pyarrow as pa
    import pyarrow.parquet as pq
    rows = [f.get("properties") or {} for f in enriched_fc.get("features", [])]
    df = pd.DataFrame(rows)
    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    table = pa.Table.from_pandas(df, preserve_index=False)
    pq.write_table(table, out_path)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--max-seconds", type=int, default=3000)
    ap.add_argument("--chunk", type=int, default=0,
                    help="limite nb munis traités ce run (0=illimité)")
    ap.add_argument("--only", help="slug unique (test)")
    ap.add_argument("--dry-run", action="store_true",
                    help="fetch+join+stats mais aucun upload")
    a = ap.parse_args()

    os.makedirs(WORK, exist_ok=True)
    os.makedirs(WORK + "/lots", exist_ok=True)
    os.makedirs(WORK + "/out", exist_ok=True)

    s3 = s3_client()
    if not os.path.exists(BOUNDARIES):
        s3.download_file(BUCKET, SDA_BOUNDARIES_KEY, BOUNDARIES)

    resolver = CodeResolver(BOUNDARIES)
    print("resolver: %d noms SDA, %d noms rôle"
          % (len(resolver.by_name_sda), len(resolver.by_name_role)))

    prog = load_prog()
    for key in ("done", "skip_exists", "no_code", "no_role", "errors"):
        prog.setdefault(key, {} if key == "done" else [])

    all_slugs = sorted(list_slugs(s3, CAD_PREFIX, ".geojson"))
    role_slugs = set(list_slugs(s3, ROLE_PREFIX, ".parquet"))

    if a.only:
        all_slugs = [s for s in all_slugs if s == a.only]
        if not all_slugs:
            print("ERREUR: slug --only=%s introuvable dans %s" % (a.only, CAD_PREFIX))
            return

    t0 = time.time()
    processed = 0
    for slug in all_slugs:
        if slug in prog["done"]:
            continue
        if slug in role_slugs:
            # parquet déjà déposé (les 30 z∩m∩p + tout commit antérieur de ce batch)
            if slug not in prog["skip_exists"]:
                prog["skip_exists"].append(slug)
            continue
        if time.time() - t0 > a.max_seconds:
            print("STOP wall-clock; relancer pour continuer")
            break
        if a.chunk and processed >= a.chunk:
            print("STOP chunk limit (%d); relancer pour continuer" % a.chunk)
            break
        processed += 1

        # 1) résoudre le code_geo
        code, method = resolver.resolve(slug)
        if code is None:
            print("NO-CODE     %-44s (skip)" % slug)
            if slug not in prog["no_code"]:
                prog["no_code"].append(slug)
            save_prog(prog)
            continue

        # 2) fetch + parse rôle
        try:
            xml = fetch_role_bytes(code, MILLESIME)
        except Exception as e:
            print("FAIL-FETCH  %-44s code=%s %s" % (slug, code, e))
            prog["errors"].append([slug, "fetch:%s" % e])
            save_prog(prog)
            continue
        if xml is None:
            print("NO-ROLE     %-44s code=%s (404, skip)" % (slug, code))
            if slug not in prog["no_role"]:
                prog["no_role"].append(slug)
            save_prog(prog)
            continue
        try:
            lookup = parse_role(xml)
        except Exception as e:
            print("FAIL-PARSE  %-44s code=%s %s" % (slug, code, e))
            prog["errors"].append([slug, "parse:%s" % e])
            save_prog(prog)
            continue

        # 3) charger la cadastre clippée + jointure
        lots = WORK + "/lots/%s.geojson" % slug
        try:
            if not os.path.exists(lots):
                s3.download_file(BUCKET, CAD_PREFIX + "%s.geojson" % slug, lots)
            with open(lots) as fh:
                lots_fc = json.load(fh)
        except Exception as e:
            print("FAIL-DL     %-44s %s" % (slug, e))
            prog["errors"].append([slug, "dl:%s" % e])
            save_prog(prog)
            _cleanup(lots)
            continue

        try:
            enriched_fc, stats = join_lots_role(lots_fc, lookup)
        except Exception as e:
            print("FAIL-JOIN   %-44s %s" % (slug, e))
            prog["errors"].append([slug, "join:%s" % e])
            save_prog(prog)
            _cleanup(lots)
            continue

        n_lots = stats["total_lots_cadastre"]
        n_match = stats["lots_matched_role"]
        join_pct = stats["coverage_pct"]
        n_matricules = len(lookup)
        print("JOIN        %-44s code=%s(%s) lots=%d matricules=%d match=%d join=%.1f%%"
              % (slug, code, method, n_lots, n_matricules, n_match, join_pct))

        if a.dry_run:
            _cleanup(lots)
            continue

        # 4) écrire parquet + upload sous la clé EXACTE slug cadastre
        out = WORK + "/out/%s.parquet" % slug
        try:
            write_parquet(enriched_fc, out)
        except Exception as e:
            print("FAIL-PARQUET %-44s %s" % (slug, e))
            prog["errors"].append([slug, "parquet:%s" % e])
            save_prog(prog)
            _cleanup(lots, out)
            continue

        upload_key = ROLE_PREFIX + "%s.parquet" % slug  # clé contrôlée, jamais _slugify
        try:
            s3.upload_file(out, BUCKET, upload_key)
        except Exception as e:
            print("FAIL-UPLOAD %-44s %s" % (slug, e))
            prog["errors"].append([slug, "upload:%s" % e])
            save_prog(prog)
            _cleanup(lots, out)
            continue

        prog["done"][slug] = {"code_geo": code, "method": method,
                              "lots": n_lots, "matricules": n_matricules,
                              "matched": n_match, "join_pct": join_pct,
                              "key": upload_key}
        save_prog(prog)
        _cleanup(lots, out)

    print("=== chunk fin : done=%d skip_exists=%d no_code=%d no_role=%d errors=%d (sur %d slugs) ==="
          % (len(prog["done"]), len(prog["skip_exists"]), len(prog["no_code"]),
             len(prog["no_role"]), len(prog["errors"]), len(all_slugs)))


def _cleanup(*paths):
    for p in paths:
        try:
            if p and os.path.exists(p):
                os.remove(p)
        except Exception:
            pass


if __name__ == "__main__":
    main()

File: acquisition/src/cadastre-role-province.ts
/**
 * cadastre-role-province.ts — Batch PROVINCE du rôle foncier : joint la cadastre
 * CLIPPÉE au rôle MAMH, en masse.
 *
 * Port fidèle de `acquisition/cadastre_role_province.py`. Parcourt les cadastres
 * clippés normalized/qc-cadastre-lots/*.geojson (~1102 munis QC) et, pour chacun,
 * télécharge + parse le rôle MAMH, le joint par matricule (= NO_LOT sans espaces)
 * et dépose un parquet enrichi sous registry/role-foncier/<slug>.parquet.
 *
 * RÉSOLUTION code_geo (sans Overpass) : slug cadastre (norm NFD) ->
 *   1. index frontières SDA local (MUS_NM_MUN -> MUS_CO_GEO)
 *   2. fallback index rôle MAMH (nom -> code_geo)
 * La clé d'upload est TOUJOURS le slug cadastre exact (jamais slugify ascii).
 *
 * IDEMPOTENT & RESUMABLE : skip si parquet existe / checkpoint
 * /tmp/role_province_progress.json ; borné par --max-seconds.
 *
 * Usage :
 *   tsx src/cadastre-role-province.ts [--max-seconds 3000] [--chunk N]
 *       [--only SLUG] [--dry-run]
 */
import { readFileSync, writeFileSync, mkdirSync, existsSync, renameSync, rmSync } from "node:fs";

import type { S3Client } from "@aws-sdk/client-s3";

import { s3Client, getBytes, putBytes, listSlugs } from "./lib/s3.js";
import { norm } from "./cadastre-clip-sda.js";
import { fetchIndex, parseRole, joinLotsRole } from "./role-foncier.js";
import { writeRoleParquet } from "./lib/parquet.js";

const PROG = "/tmp/role_province_progress.json";
const WORK = "/tmp/role_province";
const BOUNDARIES = WORK + "/qc-municipalites.geojson";

const CAD_PREFIX = "normalized/qc-cadastre-lots/";
const ROLE_PREFIX = "registry/role-foncier/";
const SDA_BOUNDARIES_KEY = "normalized/qc-admin-boundaries/qc-municipalites.geojson";

const MILLESIME = 2026;

// Alias slug cadastre -> code géo (aligné sur cadastre_clip_sda.ALIAS_SLUG_TO_CODE).
const ALIAS_SLUG_TO_CODE: Record<string, string> = {
  "eeyou-istchee-james-bay": "99060",
  "hatley-township-municipality": "45043",
};

const XML_URL = (codeGeo: string, millesime: number) =>
  `https://donneesouvertes.affmunqc.net/role/RL${codeGeo}_${millesime}.xml`;

// ---------------------------------------------------------------------------
// Checkpoint
// ---------------------------------------------------------------------------

interface DoneEntry {
  code_geo: string;
  method: string;
  lots: number;
  matricules: number;
  matched: number;
  join_pct: number;
  key: string;
}
interface Prog {
  done: Record<string, DoneEntry>;
  skip_exists: string[];
  no_code: string[];
  no_role: string[];
  errors: [string, string][];
}

function loadProg(): Prog {
  try {
    return JSON.parse(readFileSync(PROG, "utf8")) as Prog;
  } catch {
    return { done: {}, skip_exists: [], no_code: [], no_role: [], errors: [] };
  }
}
function saveProg(p: Prog): void {
  const tmp = PROG + ".tmp";
  writeFileSync(tmp, JSON.stringify(p));
  renameSync(tmp, PROG);
}

// ---------------------------------------------------------------------------
// Résolution code_geo (slug cadastre -> code géographique)
// ---------------------------------------------------------------------------

class CodeResolver {
  byNameSda: Map<string, string> = new Map();
  byNameRole: Map<string, string> = new Map();

  static async create(boundariesPath: string): Promise<CodeResolver> {
    const r = new CodeResolver();
    // 1) index frontières SDA : norm(MUS_NM_MUN) -> MUS_CO_GEO
    try {
      const g = JSON.parse(readFileSync(boundariesPath, "utf8")) as {
        features?: { properties?: Record<string, unknown> }[];
      };
      for (const f of g.features ?? []) {
        const p = f.properties ?? {};
        const code = String(p["MUS_CO_GEO"] ?? p["code"] ?? "").trim();
        const nm = String(p["MUS_NM_MUN"] ?? p["name"] ?? "");
        if (code && nm && !r.byNameSda.has(norm(nm))) r.byNameSda.set(norm(nm), code);
      }
    } catch (e) {
      console.log(`WARN SDA boundaries load failed (${String(e)})`);
    }
    // 2) index rôle MAMH : norm(nom) -> code_geo (fallback)
    try {
      const idx = await fetchIndex(MILLESIME);
      for (const e of Object.values(idx)) {
        const nm = e.nom ?? "";
        const cg = e.code_geo ?? "";
        if (nm && cg && !r.byNameRole.has(norm(nm))) r.byNameRole.set(norm(nm), cg);
      }
    } catch (e) {
      console.log(`WARN role index fetch failed (${String(e)})`);
    }
    return r;
  }

  /** Retourne [code_geo, method] ou [null, 'no-match']. */
  resolve(slug: string): [string | null, string] {
    const s = norm(slug);
    if (ALIAS_SLUG_TO_CODE[s]) return [ALIAS_SLUG_TO_CODE[s]!, "alias"];
    if (this.byNameSda.has(s)) return [this.byNameSda.get(s)!, "sda-name"];
    if (this.byNameRole.has(s)) return [this.byNameRole.get(s)!, "role-index"];
    // strip progressif des suffixes '--<mrc>' / '--2'
    const segs = (slug ?? "").split(/-{2,}/);
    for (let i = segs.length; i > 0; i--) {
      const base = norm(segs.slice(0, i).join("-"));
      if (this.byNameSda.has(base)) return [this.byNameSda.get(base)!, "sda-name-prefix"];
      if (this.byNameRole.has(base)) return [this.byNameRole.get(base)!, "role-index-prefix"];
    }
    return [null, "no-match"];
  }
}

// ---------------------------------------------------------------------------
// Rôle : fetch XML par code_geo (404/403 -> null, sans lever)
// ---------------------------------------------------------------------------

async function fetchRoleBytes(codeGeo: string, millesime = MILLESIME): Promise<Buffer | null> {
  const res = await fetch(XML_URL(codeGeo, millesime));
  if (res.status === 403 || res.status === 404) return null;
  if (!res.ok) throw new Error(`role HTTP ${res.status}`);
  return Buffer.from(await res.arrayBuffer());
}

// ---------------------------------------------------------------------------
// Parquet : écrit les properties du GeoJSON enrichi (schéma 30, sans géométrie)
// ---------------------------------------------------------------------------

async function writeParquet(
  enrichedFc: { features?: { properties?: Record<string, unknown> | null }[] },
  outPath: string,
): Promise<void> {
  const rows = (enrichedFc.features ?? []).map((f) => f.properties ?? {});
  mkdirSync(outPath.slice(0, outPath.lastIndexOf("/")), { recursive: true });
  await writeRoleParquet(rows, outPath);
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

interface Args {
  maxSeconds: number;
  chunk: number;
  only?: string;
  dryRun: boolean;
}
function parseArgs(argv: string[]): Args {
  const a: Args = { maxSeconds: 3000, chunk: 0, dryRun: false };
  for (let i = 0; i < argv.length; i++) {
    const t = argv[i]!;
    if (t === "--max-seconds") a.maxSeconds = Number(argv[++i]);
    else if (t === "--chunk") a.chunk = Number(argv[++i]);
    else if (t === "--only") a.only = argv[++i];
    else if (t === "--dry-run") a.dryRun = true;
  }
  return a;
}

function cleanup(...paths: (string | null | undefined)[]): void {
  for (const p of paths) {
    try {
      if (p && existsSync(p)) rmSync(p);
    } catch {
      /* ignore */
    }
  }
}

async function main(): Promise<void> {
  const a = parseArgs(process.argv.slice(2));

  mkdirSync(WORK, { recursive: true });
  mkdirSync(WORK + "/lots", { recursive: true });
  mkdirSync(WORK + "/out", { recursive: true });

  const s3: S3Client = s3Client();
  if (!existsSync(BOUNDARIES)) {
    writeFileSync(BOUNDARIES, await getBytes(s3, SDA_BOUNDARIES_KEY));
  }

  const resolver = await CodeResolver.create(BOUNDARIES);
  console.log(`resolver: ${resolver.byNameSda.size} noms SDA, ${resolver.byNameRole.size} noms rôle`);

  const prog = loadProg();
  prog.done ??= {};
  prog.skip_exists ??= [];
  prog.no_code ??= [];
  prog.no_role ??= [];
  prog.errors ??= [];

  let allSlugs = (await listSlugs(s3, CAD_PREFIX, ".geojson")).sort();
  const roleSlugs = new Set(await listSlugs(s3, ROLE_PREFIX, ".parquet"));

  if (a.only) {
    allSlugs = allSlugs.filter((s) => s === a.only);
    if (allSlugs.length === 0) {
      console.log(`ERREUR: slug --only=${a.only} introuvable dans ${CAD_PREFIX}`);
      return;
    }
  }

  const t0 = Date.now();
  let processed = 0;
  for (const slug of allSlugs) {
    if (slug in prog.done) continue;
    if (roleSlugs.has(slug)) {
      if (!prog.skip_exists.includes(slug)) prog.skip_exists.push(slug);
      continue;
    }
    if ((Date.now() - t0) / 1000 > a.maxSeconds) {
      console.log("STOP wall-clock; relancer pour continuer");
      break;
    }
    if (a.chunk && processed >= a.chunk) {
      console.log(`STOP chunk limit (${a.chunk}); relancer pour continuer`);
      break;
    }
    processed++;

    // 1) résoudre le code_geo
    const [code, method] = resolver.resolve(slug);
    if (code === null) {
      console.log(`NO-CODE     ${slug.padEnd(44)} (skip)`);
      if (!prog.no_code.includes(slug)) prog.no_code.push(slug);
      saveProg(prog);
      continue;
    }

    // 2) fetch + parse rôle
    let xml: Buffer | null;
    try {
      xml = await fetchRoleBytes(code, MILLESIME);
    } catch (e) {
      console.log(`FAIL-FETCH  ${slug.padEnd(44)} code=${code} ${String(e)}`);
      prog.errors.push([slug, `fetch:${String(e)}`]);
      saveProg(prog);
      continue;
    }
    if (xml === null) {
      console.log(`NO-ROLE     ${slug.padEnd(44)} code=${code} (404, skip)`);
      if (!prog.no_role.includes(slug)) prog.no_role.push(slug);
      saveProg(prog);
      continue;
    }
    let lookup: ReturnType<typeof parseRole>;
    try {
      lookup = parseRole(xml);
    } catch (e) {
      console.log(`FAIL-PARSE  ${slug.padEnd(44)} code=${code} ${String(e)}`);
      prog.errors.push([slug, `parse:${String(e)}`]);
      saveProg(prog);
      continue;
    }

    // 3) charger la cadastre clippée + jointure
    const lots = WORK + `/lots/${slug}.geojson`;
    let lotsFc: { type: string; features: { type: "Feature"; geometry: unknown; properties: Record<string, unknown> | null }[] };
    try {
      if (!existsSync(lots)) {
        writeFileSync(lots, await getBytes(s3, CAD_PREFIX + `${slug}.geojson`));
      }
      lotsFc = JSON.parse(readFileSync(lots, "utf8"));
    } catch (e) {
      console.log(`FAIL-DL     ${slug.padEnd(44)} ${String(e)}`);
      prog.errors.push([slug, `dl:${String(e)}`]);
      saveProg(prog);
      cleanup(lots);
      continue;
    }

    let enrichedFc: ReturnType<typeof joinLotsRole>["enrichedFc"];
    let stats: ReturnType<typeof joinLotsRole>["stats"];
    try {
      const r = joinLotsRole(lotsFc, lookup);
      enrichedFc = r.enrichedFc;
      stats = r.stats;
    } catch (e) {
      console.log(`FAIL-JOIN   ${slug.padEnd(44)} ${String(e)}`);
      prog.errors.push([slug, `join:${String(e)}`]);
      saveProg(prog);
      cleanup(lots);
      continue;
    }

    const nLots = stats.total_lots_cadastre;
    const nMatch = stats.lots_matched_role;
    const joinPct = stats.coverage_pct;
    const nMatricules = Object.keys(lookup).length;
    console.log(
      `JOIN        ${slug.padEnd(44)} code=${code}(${method}) lots=${nLots} ` +
        `matricules=${nMatricules} match=${nMatch} join=${joinPct.toFixed(1)}%`,
    );

    if (a.dryRun) {
      cleanup(lots);
      continue;
    }

    // 4) écrire parquet + upload sous la clé EXACTE slug cadastre
    const out = WORK + `/out/${slug}.parquet`;
    try {
      await writeParquet(enrichedFc, out);
    } catch (e) {
      console.log(`FAIL-PARQUET ${slug.padEnd(44)} ${String(e)}`);
      prog.errors.push([slug, `parquet:${String(e)}`]);
      saveProg(prog);
      cleanup(lots, out);
      continue;
    }

    const uploadKey = ROLE_PREFIX + `${slug}.parquet`; // clé contrôlée, jamais slugify
    try {
      await putBytes(s3, uploadKey, readFileSync(out), "application/octet-stream");
    } catch (e) {
      console.log(`FAIL-UPLOAD ${slug.padEnd(44)} ${String(e)}`);
      prog.errors.push([slug, `upload:${String(e)}`]);
      saveProg(prog);
      cleanup(lots, out);
      continue;
    }

    prog.done[slug] = {
      code_geo: code,
      method,
      lots: nLots,
      matricules: nMatricules,
      matched: nMatch,
      join_pct: joinPct,
      key: uploadKey,
    };
    saveProg(prog);
    cleanup(lots, out);
  }

  console.log(
    `=== chunk fin : done=${Object.keys(prog.done).length} skip_exists=${prog.skip_exists.length} ` +
      `no_code=${prog.no_code.length} no_role=${prog.no_role.length} errors=${prog.errors.length} ` +
      `(sur ${allSlugs.length} slugs) ===`,
  );
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((e) => {
    console.error(e);
    process.exit(1);
  });
}
